api/src/main/java/org/openmrs/module/reporting/common/SqlIterator.java

```diff
@@ -0,0 +1,98 @@
+package org.openmrs.module.reporting.common;
+
+import org.openmrs.module.reporting.dataset.DataSetColumn;
+import org.openmrs.module.reporting.dataset.DataSetRow;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SqlIterator implements Iterator<DataSetRow> {
+    private ResultSet resultSet;
+    private List<DataSetColumn> columns;
+    private Statement statement;
+    private Connection connection;
+    private static final Logger log = LoggerFactory.getLogger(SqlIterator.class);
+
+
+    public SqlIterator(ResultSetMetaData metadata, ResultSet resultSet, Connection connection, Statement statement) throws SQLException {
+        this.resultSet = resultSet;
+        this.connection = connection;
+        this.statement = statement;
+        this.createDataSetColumns(metadata);
+    }
+
+    @Override
+    public boolean hasNext() {
+        try {
+            if (resultSet.isLast()) {
+                closeConnection();
+                return false;
+            } else {
+                return true;
+            }
+        } catch (SQLException ex) {
+            log.error("Failed to validate last result in ResultSet.", ex);
+            return false;
+        }
+    }
+
+    @Override
+    public DataSetRow next() throws IllegalArgumentException {
+        try {
+            if (resultSet.next()) {
+                return createDataSetRow();
+            } else {
+                closeConnection();
+                return null;
```

Comment:
Did you intentionally change the contract of `Iterator:next()`? @throws NoSuchElementException if the iteration has no more elements